prov/sm2/src/sm2_msg.c

```diff
@@ -310,13 +319,14 @@ static ssize_t sm2_generic_sendmsg(struct sm2_ep *ep, const struct iovec *iov,
 	total_len = ofi_total_iov_len(iov, iov_count);
 	assert(!(op_flags & FI_INJECT) || total_len <= SM2_INJECT_SIZE);
 
-	ret = sm2_proto_ops[sm2_proto_inject](ep, peer_smr, peer_gid, op, tag,
-					      data, op_flags, mr, iov,
-					      iov_count, total_len, context);
+	proto = sm2_select_proto(total_len);
+	ret = sm2_proto_ops[proto](ep, peer_smr, peer_gid, op, tag, data,
+				   op_flags, mr, iov, iov_count, total_len,
+				   context);
 	if (ret)
 		goto unlock_cq;
 
-	if (!(op_flags & FI_DELIVERY_COMPLETE)) {
+	if (!(op_flags & FI_DELIVERY_COMPLETE) && proto != sm2_proto_cma) {
```

Comment:
Seems like by setting the op_flags |= FI_DELIVERY_COMPLETE earlier, you don't need the proto check.

Reply:
I do need the proto check b/c that op_flags is the op_flags passed in, and we modify the op_flags in the shared memory region xfer_entry.